LISP UI Refinements

dan-tang-ssd opened the pull request:
This PR is submitted to 
fix #221 
fix #222 
fix #224 

It is not yet ready for review. It is submitted for progress update.

It contains below changes:
 - [x] LISP, match with existing global indicators page, wrap global indicator in table cell
 - [ ] LISP, match with existing global indicators page, do not show empty column header
 - [ ] LISP, add custom survey questions, refine UI to show questions in tabular format, click to show details in popup modal

dave-mills commented:
The app panel has a custom theme installed, and the css is in `resources/cdd/filament/app/theme.css`

dave-mills commented:
> Is there any way to re-enable to manually add a new question directly? (It is simpler and quicker for testing)

This should be working; nothing has changed here in the last week. Do you have any local indicators that are not matched to global indicators in your database? See the blade template for that  part of the page: 

```blade
<h2 class="my-8">Add / Edit Questions Directly</h2>
@foreach($unmatchedLocalIndicators as $localIndicator)
        <livewire:local-indicator-question-form :localIndicator="$localIndicator" :key="$localIndicator->id">
@endforeach
```

dan-tang-ssd commented:
> > Is there any way to re-enable to manually add a new question directly? (It is simpler and quicker for testing)
> 
> This should be working; nothing has changed here in the last week. Do you have any local indicators that are not matched to global indicators in your database? See the blade template for that part of the page:
> 
> ```
> <h2 class="my-8">Add / Edit Questions Directly</h2>
> @foreach($unmatchedLocalIndicators as $localIndicator)
>         <livewire:local-indicator-question-form :localIndicator="$localIndicator" :key="$localIndicator->id">
> @endforeach
> ```

Oh... all of my local indicators are matched to global indicators. I just unmatched one local indicator, then I can add new question now.

![image](https://github.com/user-attachments/assets/58387140-41e8-4448-8c0b-beed54039c44)

dan-tang-ssd commented:
I did a CSS hack for the 2nd issue. But it also affects other tables in other pages, e.g. Survey Dashboard > Place Adaptations > Diet Diversity.

Question:
1. Is there any way to add something to a table to make it uniquely identified in the whole application, so that we can do a CSS hack to that particular table only?

---

Screen shots:

BEFORE:
![image](https://github.com/user-attachments/assets/daf6728a-00df-4e86-8b41-8761d337b3b2)

AFTER:
![image](https://github.com/user-attachments/assets/a3e54ac5-4d25-4460-89ae-6684a1ee005f)

BEFORE:
![image](https://github.com/user-attachments/assets/a96cb507-5959-42ae-ac29-0988b461578a)

AFTER:
![image](https://github.com/user-attachments/assets/49831847-68ee-4ff0-956c-367c379fecfb)

dave-mills commented:
There are lots of options for making your css only affect specific targets. I'd just add a class to a parent element and make your css target only children of that class.

dan-tang-ssd commented:
Thank you @dave-mills for your advice.

I added a CSS class "lisp-global-indicators" by adding a "div" tag to the global indicators table.
Now the CSS hack takes effect on this specific table only.

dan-tang-ssd commented:
@dave-mills - progress update, and questionsss...

1. I added a table() to return SurveyRow records. Not sure how to do "DRAG TO ORDER" as reorderableWithDragAndDrop() and reorderAction() belong to Repeater. In this case, should I use Repeater in a table?
2. I filled the modal popup form with existing data, but changes for type and name cannot be saved.
3. I added a "ADD QUESTION" button in table header, how to connect it to livewire component to call related function to add a question? (or maybe just show a blank popup modal form?)

P.S. I keep the original form visible for quick checking. I will remove it once the new table way working properly.

---

Screen shots:

![image](https://github.com/user-attachments/assets/7a7ab170-4a67-4c26-b766-b06423542d11)

![image](https://github.com/user-attachments/assets/9a326072-4c9a-40d4-9125-d7034a4d1707)

dan-tang-ssd commented:
Thank you @dave-mills for your advice.

I think it is now ready for review.

The "Edit Question" and "Add Question" button use two separate forms with same content. We may consolidate the forms into one form as further improvement.

---

Screen shots:

![image](https://github.com/user-attachments/assets/c89c3e26-0f02-486a-a133-01041de41cf4)

![image](https://github.com/user-attachments/assets/a9dec355-d0ee-4abf-8c78-cc303e4a3012)

![image](https://github.com/user-attachments/assets/b59c7b32-dd79-4fa7-a545-41459728720b)

dave-mills reviewed (commented):
Looks good. The change to a table view with the "Question details"  form inside a modal is definitely much better. Reordering works well. 

Couple of things to look at: 

- When creating a new question: A ChoiceList entry is created for select questions, but the list_name is "_choices", instead of "{$questionName}_choices". Also, no ChoiceListEntry entries are created.  Not sure why, as the forms look the same in both create and edit, and both things work fine in the edit form.

- The "default label" in the table currently shows the JSON representation of the whole LanguageString. (`defaultLabel` and `defaultHint` were recently changed to be a HasOne relationship instead of an attribute that calculates the `text`).

dave-mills commented on app/Livewire/LocalIndicatorQuestionForm.php:
```diff
@@ -92,25 +99,49 @@ public function form(Form $form): Form
+                    ->mutateFormDataUsing(function (array $data): array {
+                        // find the largest row_number of survey_rows records
+                        $xlsformModuleVersion = XlsformModuleVersion::find($this->xlsformModuleVersion->id);
+                        $lastRowNumber = $xlsformModuleVersion->surveyRows == null ? $xlsformModuleVersion->surveyRows->last()->row_number : 0;
```
Is this logic the wrong way round? All my questions are getting a `row_number` of 1 until I manually re-order them.

dan-tang-ssd replied:
Just fixed this logical error, the condition should be != null instead of == null

dan-tang-ssd commented on app/Livewire/LocalIndicatorQuestionForm.php:
```diff
@@ -92,25 +99,49 @@ public function form(Form $form): Form
+                    ->createAnother(false)
+                    ->mutateFormDataUsing(function (array $data): array {
+                        // find the largest row_number of survey_rows records
+                        $xlsformModuleVersion = XlsformModuleVersion::find($this->xlsformModuleVersion->id);
```
Er.... YES!

dave-mills commented:
Ah yes - you're right! I hadn't updated the submodule so I was seeing incorrect results. You're right - we'll update it after merging in the big ODK update.

dave-mills reviewed (approved):
Looks good to me!

dan-tang-ssd commented:
Resolved merge conflicts in Github web site.
I will merge this PR into dev branch now.
